test/gtest/ucp/test_ucp_proto.cc

```diff
@@ -431,16 +441,118 @@ class test_proto_perf : public ucs::test {
+        return range->start;
+    }
+
+    template<typename PerfPtrType>
+    void expect_perf_empty_range(PerfPtrType perf, size_t start, size_t end)
+    {
+        ASSERT_NE(perf.get(), nullptr);
+        auto *seg = find_lb(perf, start);
+
+        EXPECT_TRUE((seg == nullptr) || (end < get_start(seg)))
+                << "perf=" << perf.get() << " start=" << start
                 << " end=" << end;
     }
 
+    void expect_empty_range(size_t start, size_t end)
+    {
+        expect_perf_empty_range(m_perf, start, end);
+        expect_perf_empty_range(m_sum_flat_perf, start, end);
+        expect_perf_empty_range(m_envelope_flat_perf, start, end);
+    }
+
+    void expect_flat_range(flat_perf_ptr_t flat_perf, size_t start, size_t end,
+                           ucs_linear_func_t exp_func)
+    {
+        auto *range     = find_lb(flat_perf, start);
+        size_t midpoint = start + (end - start) / 2;
+
+        std::stringstream info;
+        info << "expected=[" << start << ", " << end << ", " << exp_func
+             << "] actual=[" << range->start << ", " << range->end << ", "
+             << range->value << "]";
+
+        EXPECT_NE(range, nullptr) << info.str();
+        EXPECT_GE(start, range->start) << info.str();
+        EXPECT_LE(end, range->end) << info.str();
+
+        // Cannot compare functions directly since on big sizes (e.g. SIZE_MAX)
+        // small latency difference can be overhelmed by floating point math
+        // innacurracy after addition with huge BW*SIZE_MAX.
+        for (size_t point: {start, midpoint, end}) {
+            double expected_result = ucs_linear_func_apply(exp_func, point);
+            double range_result    = ucs_linear_func_apply(range->value, point);
+            EXPECT_NEAR(expected_result, range_result, 10e-9) << info.str();
+        }
+    }
+
+    void expect_envelope_flat_perf(size_t start, size_t end,
+                                   const perf_factors_map_t &factors)
+    {
+        ucp_proto_perf_envelope_elem_t *envelope_elem;
+        std::vector<ucs_linear_func_t> factors_vec;
+        ucp_proto_perf_envelope_t envelope;
+
+        for (const auto& factor: factors) {
+            if (factor.first == UCP_PROTO_PERF_FACTOR_LATENCY) {
+                continue;
+            }
+            factors_vec.emplace_back(factor.second);
+        }
+
+        ucs_array_init_dynamic(&envelope);
+        ASSERT_UCS_OK(ucp_proto_perf_envelope_make(
+                &factors_vec.front(), UCS_MASK(factors_vec.size()), start, end,
+                0, &envelope));
+
+        ucs_array_for_each(envelope_elem, &envelope) {
+            expect_flat_range(m_envelope_flat_perf, start,
+                              envelope_elem->max_length,
+                              factors_vec[envelope_elem->index]);
+
+            start = envelope_elem->max_length + 1;
+        }
+
+        ucs_array_cleanup_dynamic(&envelope);
+    }
+
+    void expect_sum_flat_perf(size_t start, size_t end,
```

Comment:
let's move all function that are not called directly from TEST functions to be private (instead of protected)

Reply:
Done.